backend/src/services/hsm/hsm-service.ts

```diff
@@ -0,0 +1,349 @@
+      const { wrappedKey, sessionKey } = $generateAndWrapKey(s);
+
+      const alg = {
+        name: appCfg.HSM_MECHANISM,
+        params: new graphene.AesGcm240Params(iv)
+      } as grapheneLib.IAlgorithm;
+
+      const cipher = s.createCipher(alg, new graphene.Key(sessionKey).toType());
+
+      // Calculate the output buffer size based on input length
+      // GCM adds a 16-byte auth tag, so we need input length + 16
+      const outputBuffer = Buffer.alloc(data.length + TAG_LENGTH);
+      const encryptedData = cipher.once(data, outputBuffer);
+
+      // Format: [Wrapped Key (40)][IV (16)][Encrypted Data + Tag]
+      return Buffer.concat([wrappedKey, iv, encryptedData]);
+    };
+
+    if (providedSession) {
+      return $performEncryption(providedSession);
+    }
+
+    const encrypted = await withSession({ module, graphene }, $performEncryption);
+
+    return encrypted;
+  };
+
+  const decrypt: {
+    (encryptedBlob: Buffer, providedSession: grapheneLib.Session): Promise<Buffer>;
+    (encryptedBlob: Buffer): Promise<Buffer>;
+  } = async (encryptedBlob: Buffer, providedSession?: grapheneLib.Session) => {
+    if (!module) {
+      throw new Error("HSM service not initialized");
+    }
+
+    const $performDecryption = (s: grapheneLib.Session) => {
+      const wrappedKey = encryptedBlob.subarray(0, WRAPPED_KEY_LENGTH);
+      const iv = encryptedBlob.subarray(WRAPPED_KEY_LENGTH, WRAPPED_KEY_LENGTH + IV_LENGTH);
+      const ciphertext = encryptedBlob.subarray(WRAPPED_KEY_LENGTH + IV_LENGTH);
+
+      // Unwrap the session key
+      const sessionKey = $unwrapKey(s, wrappedKey);
+
+      const algo = {
+        name: appCfg.HSM_MECHANISM,
+        params: new graphene.AesGcm240Params(iv)
+      };
+
+      const decipher = s.createDecipher(algo, new graphene.Key(sessionKey).toType());
+      const outputBuffer = Buffer.alloc(ciphertext.length);
+
+      // Extract wrapped key, IV, and ciphertext
+      return decipher.once(ciphertext, outputBuffer);
+    };
+
+    if (providedSession) {
+      return $performDecryption(providedSession);
+    }
+    const decrypted = await withSession({ module, graphene }, (newSession) => $performDecryption(newSession));
+
+    return decrypted;
+  };
+
+  // We test the core functionality of the PKCS#11 module that we are using throughout Infisical. This is to ensure that the user doesn't configure a faulty or unsupported HSM device.
+  const $testPkcs11Module = async (session: grapheneLib.Session) => {
+    try {
+      if (!module) {
+        throw new Error("HSM service not initialized");
+      }
+
+      if (!session) {
+        throw new Error("Session not initialized");
+      }
+
+      const randomData = session.generateRandom(256);
+      const encryptedData = await encrypt(Buffer.from(randomData), session);
+      const decryptedData = await decrypt(encryptedData, session);
+
+      if (Buffer.from(randomData).toString("hex") !== Buffer.from(decryptedData).toString("hex")) {
```

Comment:
Nit: I think can use `Buffer.equals()` to compare the buffers instead of needing to do it via conversion to hex.